src/utils/actions.js

```diff
@@ -0,0 +1,396 @@
+  // information about applicability conditions
+  // TODO: Does the applicability condition need
+  // to differentiate between patient/practitioner/encounter?
+  // should the operating resources be put in the context separately?
+
+  const planDef = context.planDef;
+  if (planDef.condition) {
+    // assume that if the condition is satisfied by one of the three
+    // then it is satisfied
+    const patient = context.patient;
+    const encounter = context.encounter;
+    const resources = [patient, encounter];
+    const boolMap = planDef.condition.map(condition => {
+      if (condition.kind === 'applicability' && condition.expression) {
+        const expression = condition.expression;
+        // does the condition need to be true for any or all resources?
+        return resources.some(resource => {
+          return evaluateExpression(expression, resource);
+        });
+      } else {
+        return true;
+      }
+    });
+  }
+}
+
+function createReport(context) {
+  context.contentBundle = createBundle(context.records);
+  const reportingBundle = createBundle(context.records);
+  // assume the context has a messageHeader.
+  reportingBundle.entry.unshift(context.messageHeader);
+  context.reportingBundle = reportingBundle;
+  return context;
+}
+
+function validateReport(context) {
+  // assume that we're getting the validation
+  // output from action.output.profile
+  const isValid = true;
+
+  action.output.forEach(output => {
+    const profile = output.profile;
+    if (profile) {
+      const canonicalUrl = profile.reference;
+      // assume the client can read the canonicalUrl
+      const structureDefinition = context.client.source.read(canonicalUrl); // will probably need to deal with a Promise
+      // if there's an invalid, the whole thing is invalid
+      isValid = validateProfile(context.reportingBundle, structureDefinition) & isValid;
+    }
+  });
+
+  if (isValid) {
+    context.flags['valid'] = true;
+  } else {
+    context.flags['valid'] = false;
+  }
+}
+
+function submitReport(context) {
+  context.client.dest.submit(context.reportingBundle);
+}
+
+function deidentifyReport(context) {
+  const client = context.client.trustServices;
+  client.deidentify(context.contentBundle);
+}
+
+function anonymizeReport(context) {
+  const client = context.client.trustServices;
+  client.anonymize(context.contentBundle);
+}
+
+function pseudonymizeReport(context) {
+  const client = context.client.trustServices;
+  client.pseudonymize(context.contentBundle);
+}
+
+function encryptReport(context) {
+  // TODO: Are we meant to actually encrypt the bundle?
+  // do we need an external library to encrypt the
+  // stringified bundle?
+  const bundle = context.reportingBundle;
+  const base64 = btoa(JSON.stringify(bundle));
+  const resource = {
+    resourceType: 'Binary',
+    type: 'application/json',
+    data: base64
+  };
+  context.encryptedBundle = resource;
```

Comment:
The way I envisioned this is a flow from one of these actions to another. Each one is going to manipulate the report until its submitted and the workflow is done. As such I don't think these should all be saved as different types of bundles in the context. There's no point encrypting it and saving it to `context.encryptedBundle` if no other action uses that (and since an action will not know which other actions have been called previously you can't guarantee this property has been set). So I think all of these should be saving the updated bundle to the same property of context.